File: backend/module1_learner/extractors/state_machine.py
#!/usr/bin/env python3
"""
State Machine Extractor
Identifies state variables and their transitions from logs
"""

import re

class StateMachineExtractor:
    def __init__(self):
        self.state_variables = {}
        self.transitions = []
        
    def extract(self, logs):
        """Extract state machines from logs"""
        print("   - Analyzing state changes...")
        
        # Look for common state patterns
        state_patterns = {
            'counters': self._find_counters(logs),
            'flags': self._find_flags(logs),
            'sequences': self._find_sequences(logs),
            'values': self._find_tracking_values(logs)
        }
        
        return state_patterns
    
    def _find_counters(self, logs):
        """Find variables that increment (like step counts)"""
        counters = {}
        last_values = {}
        
        for log in logs:
            if 'message' not in log:
                continue
                
            msg = log['message']
            
            # Look for patterns like "onStandStepChanged 3579"
            counter_matches = re.findall(r'(\w+)\s+(\d+)', msg)
            
            for name, value in counter_matches:
                value = int(value)
                if name not in last_values:
                    last_values[name] = value
                    counters[name] = {'type': 'counter', 'samples': [value]}
                else:
                    if value > last_values[name]:
                        counters[name]['type'] = 'incrementing_counter'
                    counters[name]['samples'].append(value)
                    last_values[name] = value
        
        return counters
    
    def _find_flags(self, logs):
        """Find boolean state variables"""
        flags = {}
        flag_patterns = ['SCREEN_ON', 'SCREEN_OFF', 'true', 'false']
        
        for log in logs:
            if 'message' not in log:
                continue
                
            msg = log['message']
            for pattern in flag_patterns:
                if pattern in msg:
                    flags[pattern] = flags.get(pattern, 0) + 1
        
        return flags
    
    def _find_sequences(self, logs):
        """Find sequential state changes"""
        sequences = []
        last_state = None
        
        for log in logs:
            if 'message' not in log:
                continue
                
            current = log['message'][:30]  # First 30 chars as rough state
            
            if last_state and last_state != current:
                sequences.append({
                    'from': last_state,
                    'to': current,
                    'component': log.get('component', 'unknown')
                })
            
            last_state = current
        
        # Find common transitions
        transition_counts = {}
        for seq in sequences:
            key = f"{seq['from']} → {seq['to']}"
            transition_counts[key] = transition_counts.get(key, 0) + 1
        
        return {
            'common_transitions': dict(sorted(transition_counts.items(), 
                                             key=lambda x: x[1], reverse=True)[:20]),
            'total_sequences': len(sequences)
        }
    
    def _find_tracking_values(self, logs):
        """Find values that are tracked over time"""
        tracked = {}
        
        value_patterns = [
            (r'totalSteps=(\d+)', 'totalSteps'),
            (r'totalCalories=(\d+)', 'totalCalories'),
            (r'totalAltitude=(\d+)', 'totalAltitude'),
            (r'patient_id=(\d+)', 'patient_id')
        ]
        
        for log in logs:
            if 'message' not in log:
                continue
                
            msg = log['message']
            for pattern, name in value_patterns:
                match = re.search(pattern, msg)
                if match:
                    if name not in tracked:
                        tracked[name] = []
                    tracked[name].append(int(match.group(1)))
        
        # Calculate ranges
        for name, values in tracked.items():
            if values:
                tracked[name] = {
                    'min': min(values),
                    'max': max(values),
                    'avg': sum(values) / len(values),
                    'samples': len(values)
                }
        
        return tracked

File: log-collector/internal/learner/state_machine.go
package learner

import (
	"regexp"
	"sort"
	"strconv"
	"strings"

	"migrate/internal/models"
)

// StateMachineExtractor extracts state machines from logs
type StateMachineExtractor struct{}

// NewStateMachineExtractor creates a new state machine extractor
func NewStateMachineExtractor() *StateMachineExtractor {
	return &StateMachineExtractor{}
}

// Extract extracts state machine patterns
func (sme *StateMachineExtractor) Extract(logs []models.LogEntry) models.StateMachine {
	return models.StateMachine{
		Counters:  sme.findCounters(logs),
		Flags:     sme.findFlags(logs),
		Sequences: sme.findSequences(logs),
		Values:    sme.findTrackingValues(logs),
	}
}

// findCounters finds incrementing counters
func (sme *StateMachineExtractor) findCounters(logs []models.LogEntry) map[string]models.Counter {
	counters := make(map[string]models.Counter)
	lastValues := make(map[string]int)
	
	for _, log := range logs {
		message, ok := log.Fields["message"].(string)
		if !ok {
			continue
		}
		
		// Look for patterns like "word number"
		re := regexp.MustCompile(`(\w+)\s+(\d+)`)
		matches := re.FindAllStringSubmatch(message, -1)
		
		for _, match := range matches {
			name := match[1]
			value, _ := strconv.Atoi(match[2])
			
			if lastVal, exists := lastValues[name]; exists {
				counter := counters[name]
				counter.Samples = append(counter.Samples, value)
				if value > lastVal {
					counter.Type = "incrementing_counter"
				}
				counters[name] = counter
			} else {
				counters[name] = models.Counter{
					Type:    "counter",
					Samples: []int{value},
				}
			}
			lastValues[name] = value
		}
	}
	
	return counters
}

// findFlags finds boolean state variables
func (sme *StateMachineExtractor) findFlags(logs []models.LogEntry) map[string]int {
	flags := make(map[string]int)
	flagPatterns := []string{"SCREEN_ON", "SCREEN_OFF", "true", "false", "ON", "OFF"}
	
	for _, log := range logs {
		message, ok := log.Fields["message"].(string)
		if !ok {
			continue
		}
		
		for _, pattern := range flagPatterns {
			if strings.Contains(message, pattern) {
				flags[pattern]++
			}
		}
	}
	
	return flags
}

// findSequences finds sequential state changes
func (sme *StateMachineExtractor) findSequences(logs []models.LogEntry) models.SequenceInfo {
	type transition struct {
		from string
		to   string
	}
	
	sequences := make([]transition, 0)
	
	var lastMessage string
	
	for _, log := range logs {
		message, ok := log.Fields["message"].(string)
		if !ok {
			continue
		}
		
		// Use first 30 chars as state
		current := message
		if len(current) > 30 {
			current = current[:30]
		}
		
		if lastMessage != "" && lastMessage != current {
			sequences = append(sequences, transition{lastMessage, current})
		}
		
		lastMessage = current
	}
	
	// Count transitions
	transitionCounts := make(map[string]int)
	for _, seq := range sequences {
		key := seq.from + " → " + seq.to
		transitionCounts[key]++
	}
	
	// Sort and get top 20
	sortedTransitions := make(map[string]int)
	type kv struct {
		Key   string
		Value int
	}
	var sorted []kv
	for k, v := range transitionCounts {
		sorted = append(sorted, kv{k, v})
	}
	
	// Sort descending
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Value > sorted[j].Value
	})
	
	for i, kv := range sorted {
		if i >= 20 {
			break
		}
		sortedTransitions[kv.Key] = kv.Value
	}
	
	return models.SequenceInfo{
		CommonTransitions: sortedTransitions,
		TotalSequences:    len(sequences),
	}
}

// findTrackingValues finds values that are tracked over time
func (sme *StateMachineExtractor) findTrackingValues(logs []models.LogEntry) map[string]models.ValueInfo {
	tracked := make(map[string][]int)
	
	valuePatterns := []struct {
		pattern string
		name    string
	}{
		{`totalSteps=(\d+)`, "totalSteps"},
		{`totalCalories=(\d+)`, "totalCalories"},
		{`totalAltitude=(\d+)`, "totalAltitude"},
		{`patient_id=(\d+)`, "patient_id"},
		{`step_count[=:\s]+(\d+)`, "step_count"},
	}
	
	for _, log := range logs {
		message, ok := log.Fields["message"].(string)
		if !ok {
			continue
		}
		
		for _, vp := range valuePatterns {
			re := regexp.MustCompile(vp.pattern)
			if match := re.FindStringSubmatch(message); len(match) > 1 {
				val, _ := strconv.Atoi(match[1])
				tracked[vp.name] = append(tracked[vp.name], val)
			}
		}
	}
	
	// Calculate statistics
	result := make(map[string]models.ValueInfo)
	for name, values := range tracked {
		if len(values) > 0 {
			min, max, sum := values[0], values[0], 0
			for _, v := range values {
				if v < min {
					min = v
				}
				if v > max {
					max = v
				}
				sum += v
			}
			
			result[name] = models.ValueInfo{
				Min:     min,
				Max:     max,
				Avg:     float64(sum) / float64(len(values)),
				Samples: len(values),
			}
		}
	}
	
	return result
}
